### DISCUSSION TITLE
Beta 2.0.8(20230512)

### DISCUSSION THREAD

**Discussion by bao-cn** [Lunova-Studio/DarkMode2#13] [Polls]
Add:
1. Wallpaper Engine Support
2. User information collection
3. Improve related pages
Fix:
1. Startup
2. System version control

<hr /><em>This discussion was created from the release <a href='https://github.com/Melon-Studio/DarkMode2/releases/tag/2.0.8.20230512-Beta'>Beta 2.0.8(20230512)</a>.</em>

**Comment by orency** [comment]
一点设置就闪退了😅没法点到重置用户配置啊!请问可以手动删除配置文件让他恢复正常吗?用户配置文件位置在哪儿呢?

**Reply by bao-cn** [reply]
*注意：在使用Beta版本时，请清除注册表键值，Beta版本处于快速迭代开发状态，很多功能都在变动。*

清除方式：

 1. 在设置中心的设置中，点击最下方`重置用户配置`按钮；
 2. 在 `Windows 搜索`中搜索`注册表编辑器`，打开后在地址栏输入`计算机\HKEY_CURRENT_USER\Software\DarkMode2`，将这个`DarkMode2`项删除。

**Reply by orency** [reply]
感谢😊问题解决了.
提供2点建议
1.添加个手动输入经纬度或根据ip网络定位的功能,因为绝大部分win设备没有gps传感器,windows位置服务基本是个残废呵!
2.根据经纬度算出的日出日落时间其实并不是切换的最佳时机呵!因为春夏季其实日出前半小时到一小时在曙光阶段天就已经全亮了,日落后大概半小时暮光阶段亮度还是可以的,而秋冬季正好反之.所以需要由用户设置偏移一定时间的或直接算出当天的曙光暮光时间供用户选择呵

**Reply by bao-cn** [reply]
手动输入经纬度或根据ip网络定位的功能可以存在，感谢建议。

由用户设置偏移一定时间的功能将不会去设计。

通过经纬度计算日出日落时间是基于太阳在天空中的位置的变化。具体来说，日出时太阳刚好从地平线上升起，日落时太阳刚好到达地平线下方。我们需要知道太阳在天空中的位置以及地球的几何特征。

首先，黄赤交角是地球自转轴与公转轨道法线的夹角。它的值是23.5度左右，决定了太阳在天空中的最高和最低位置。其次，太阳赤尾正弦值表示太阳在天空中的纬度位置。球面三角学公式可以用来计算太阳在天空中的高度角和方位角，从而确定太阳在天空中的位置。

根据地球自转轴的倾斜和地球绕太阳的不规则椭圆轨道，太阳的平均近点角和近日点的平均黄道经度都会随时间而变化。同时，地球的椭圆公转轨道离心率也会影响太阳在天空中的位置。这些参数的变化会导致太阳的视位置和日出日落时间发生变化。

不考虑SPA算法和NREL算法，给你举个例子：

假设我们现在要计算北京市（东经116.4度，北纬39.9度）2023年5月18日的日出日落时间。

首先，我们需要确定当日的太阳平均近点角，可以通过以下公式计算：


M = 357.5291 + 0.98560028 * (J - 2451545)


其中J为儒略日，J=2459345.712将对应2023年5月18日的中午12点整。

通过这个公式计算得到当日的M为36.56度。

接下来，我们需要计算离心率e，可以通过以下公式计算：


e = 0.016708634 - 0.000042037 * T - 0.0000001267 * T^2


其中T为儒略世纪数，T=(J-2451545)/36525

带入公式计算得到e为0.016741。

然后，我们需要计算黄道经度L和太阳的偏近点角G，这可以通过以下公式计算：


L = (M + 1.9148 * sin(G) + 0.0200 * sin(2G) + 282.634) mod 360


G = (357.528 + 0.9856003 * (J - 2451545) + 1.915 * sin(M * pi / 180) + 0.020 * sin(2 * M * pi / 180)) mod 360


带入公式计算得到L为67.52度，G为26.36度。

接下来，我们需要计算太阳的赤纬，可以通过以下公式计算：


sinδ = sinε * sinβ + cosε * cosβ * cosL

cosδ = sqrt(1 - sin^2 δ)


其中ε为黄赤交角，已知为23.44度左右。β为太阳的纬度，可以通过以下公式计算：


sinβ = sinL * sinε

cosβ = (cos^2 L + (1 - e^2) * sin^2 L)^0.5 * cosε / cosδ


带入公式计算得到当日的太阳赤纬为20.17度。

最后，我们需要计算日出和日落的时分。这个计算需要考虑观测位置的经纬度和太阳的位置角。具体来说，日出时太阳的位置角为东方，日落时太阳的位置角为西方。可以通过以下公式计算：


cosω = (sin(-0.8333pi/180) - sinφsinδ) / (cosφ*cosδ)

t = 1/15 * acos(ω) * (180/pi)


其中，φ为观测点的纬度，已知为39.9度，ω为时角余弦，t为时差，即日出时间或日落时间与当地标准时间的差值（单位为小时）。对于北京市（东8区），当地时间比UTC时间快8小时。

带入公式计算得到当日的日出时间为06:04（当地时间），日落时间为19:24（当地时间）。

通过经纬度计算出的日出日落时间并没有考虑到太阳光传播到地球所需的时间（也称为光行时），就是你说的曙光阶段，因此与实际观测到的时间可能存在一定的偏差。太阳光在传输到地球表面的过程中会受到大气折射和散射的影响，使得太阳的视位置会有所改变，从而影响日出和日落时间的精度。而这种影响因地而异，取决于观测点的地理位置和当日大气条件的不同。

如果需要计算更精确的日出日落时间，还需要进一步考虑大气折射、散射以及天文学中的一些细节，如地理纬度、方位角、地球半径、大气折射率等因素。同时，还需要对当天的大气条件有相应的了解，才能更准确地计算出光行时和太阳的真实位置，得到更准确的日出日落时间。

DarkMode完全没有必要非常精确的去计算日出日落时间，用户也不会具体感知刚日出日落对于设备的颜色模式的具体影响。所以不会多此一举由用户设置偏移时间。

**Reply by orency** [reply]
曙光暮光都是太阳仍在地平线下,但由于大气的折射,所以部分光线已经到达该地,在天文网站都是可以查询到具体时间的,表明也是可以算出来的
日出日落立即切换用户是可以感知到时机不恰当不舒服的,因为这个偏移并不是光线到地球的六分钟呵!起码半小时到一小时呵.日出时可能没多大影响毕竟很少有起早使用设备的,但日落时仍在使用的用户还是挺多的
当然你这是舍本求末了😅反正偏移由用户提供,所以并不需要复杂的计算呵!同一地在一两个月内的偏移时间基本差不了多少,顶多是夏季偏日出前一小时左右,偏日落后半小时左右.冬季反之
而且就同有些人全暗了才要开灯,有些半暗就已经受不了了,不同人对光线的感知也是不同的(例如正常人觉得正好的亮度近视者会嫌暗远视者会嫌亮).所以也不应该由计算得出呵.反正就是最终结果加减一个数值的事,也并没有太复杂

**Reply by bao-cn** [reply]
这个功能还需商榷再定，目前在开发第三种切换方式，将支持感光元件的电脑根据周围亮度自适应改变颜色模式

**Comment by lyy33441314** [comment]
你好问一下 我手动删除之后 这个 开机自启 不生效 有一样问题的吗？

**Reply by lyy33441314** [reply]
这个开机不生效问题,我回退到上个版本也不行了。 只要更新了 这个版本 就不行了

**Reply by bao-cn** [reply]
开机自启功能正在改，以后会通过DarkMode 服务自启，请期待未来版本

**Reply by SHNN16** [reply]
现在通过任务计划程序启动我遇到了一个问题，自动设置的任务计划出发条件是系统启动时，没有效果，我改成了登录时就可以
